client.go

```diff
@@ -542,19 +545,26 @@ func (c *Client) Do(req *Request) (*http.Response, error) {
 		}
 	}
 
-	if c.ErrorHandler != nil {
+	defer func() {
 		c.HTTPClient.CloseIdleConnections()
-		return c.ErrorHandler(resp, err, c.RetryMax+1)
+	}()
+
+	if err == nil {
+		return resp, nil
+	}
+
+	if c.ErrorHandler != nil {
+		return c.ErrorHandler(resp, err, attempt)
 	}
 
 	// By default, we close the response body and return an error without
 	// returning the response
 	if resp != nil {
-		resp.Body.Close()
+		c.drainBody(resp.Body)
 	}
-	c.HTTPClient.CloseIdleConnections()
-	return nil, fmt.Errorf("%s %s giving up after %d attempts",
-		req.Method, req.URL, c.RetryMax+1)
+
+	return nil, fmt.Errorf("%s %s giving up after %d attempt(s): %v",
```

Comment:
This seems like a good candidate to change to a `%w` - that way the underlying error can be unwrapped if desired. https://blog.golang.org/go1.13-errors

Reply:
Is `%w` compatible with 1.12?

Reply:
Hm, perhaps not. Is that an issue? Not a huge deal and we can punt on doing %w for now, but it would be nice to get the additional error if possible.

Reply:
Not sure if it's an issue. May just want to call out in readme the go version required.

Reply:
We can declare it in the `go.mod` file at a minimum. We could also prevent people from building this package on < go1.13 as well. That might be a bit heavy-handed.

Reply:
What are your thoughts on that?

Reply:
Oh, nevermind. Looks like it's already there. So maybe nothing to do here, and we can make use of `%w`.